### ISSUE TITLE
OPTIONS on file URL should always return 204

### ISSUE THREAD

**Issue by KieranP** [tus/tus-resumable-upload-protocol#102]
Backstory:
https://github.com/tus/tus-js-client/issues/50
https://github.com/janko-m/tus-ruby-server/issues/2

Basically, if server returns 404 on options request for a file url because the file isn't there, then the browser refuses to try a HEAD command for security reasons, resulting in failure of any ability to upload/resume the same file, unless you clear local storage.

Recommend changing the spec to have something like "If OPTIONS is implemented, server MUST return a '204 No Content'"

**Comment by janko-m** [comment]
I cannot find the CORS specification describing the "preflight" OPTIONS request, but I found this in the MDN docs:

> Unlike simple requests (discussed above), "preflighted" requests first send an HTTP request by the OPTIONS method to the resource on the other domain, in order to determine whether the actual request is **safe to send**.

I think this means that each valid OPTIONS request *should* return a successful response, by the definition on CORS specification itself. An additional reason why tus-ruby-server shouldn't return 404 for a request on an non-existing file is that it wouldn't differ from a 404 that the server would return on a preflight request for a path that doesn't exist (and in latter case the preflight request should fail).

So I think this only a problem with the tus-ruby-server not following the CORS specification properly (which I'll fix shortly). Though it could also be mentioned in the tus specification, in case people missed this part when reading the CORS specification.

**Comment by Acconut** [comment]
> Recommend changing the spec to have something like "If OPTIONS is implemented, server MUST return a '204 No Content'"

Just a minor comment to your choice of the status code. While 204 seems to be the most appropriate one, some older browsers only accept 200 responses for an OPTIONS request, see https://github.com/tus/tusd/blob/master/unrouted_handler.go#L166-L173.

> An additional reason why tus-ruby-server shouldn't return 404 for a request on an non-existing file is that it wouldn't differ from a 404 that the server would return on a preflight request for a path that doesn't exist (and in latter case the preflight request should fail).

I also did look in various places for recommendations on how to handle an OPTIONS request to an URL which represents a non-existing resource but did not find anything relevant. Therefore I am not really a fan of restricting the allowed status codes to only one because we can't be sure that this is the right way to handle these requests, for now.

> So I think this only a problem with the tus-ruby-server not following the CORS specification properly (which I'll fix shortly). Though it could also be mentioned in the tus specification, in case people missed this part when reading the CORS specification.

I agree, that this is more an implementation detail and therefore does not necessary fit into the specification. We are working on a separate guide about the tus protocol which explains certain decision we made and could also feature a list of topics to consider while building a client or server. This would be a better place for this section, IMHO.

Furthermore, as I pointed out in my comment in https://github.com/tus/tus-js-client/issues/50#issuecomment-261715544, this issue would also have been fixed by using the `retryDelays` setting for the tus-js-client.
